blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/EnumValidator.java

```diff
@@ -0,0 +1,146 @@
+    }
+
+    private void validateThresholdAndWrite(Locator locator, ArrayList<String> currentEnumValues) {
+        // check if count of current enum values for the metric exceed a configurable threshold number
+        log.debug(String.format("EnumValidator validating locator %s", locator.toString()));
+
+        // if exceeded, mark metric as bad, else index enum values in elasticsearch
+        if ((currentEnumValues != null) && (currentEnumValues.size() > ENUM_UNIQUE_VALUES_THRESHOLD)) {
+            // count of current enum values of metric exceeded threshold, bad metric
+            // write locator to bad metric table
+            try {
+                AstyanaxWriter.getInstance().writeExcessEnumMetric(locator);
+            } catch (ConnectionException e) {
+                log.error(String.format("Exception writing bad metric %s: %s", locator.toString(), e.getMessage()), e);
+            }
+        }
+        else {
+            // not bad metric, create or update enums index of metric in elasticsearch if different from cassandra
+            // search for metric from elastic search
+            List<SearchResult> esSearchResult = null;
+            try {
+                esSearchResult = discoveryIO.search(locator.getTenantId(), locator.getMetricName());
+            }
+            catch (Exception e) {
+                log.error(String.format("Exception retrieving enum values from elasticsearch for %s: %s", locator.toString(), e.getMessage()), e);
+            }
+
+            // get elasticsearch enum values from top search results of exact match
+            ArrayList<String> elasticsearchEnumValues = null;
+            if ((esSearchResult != null) && (esSearchResult.get(0) != null)) {
+                elasticsearchEnumValues = esSearchResult.get(0).getEnumValues();
+            }
+
+            // sort the two array lists of enum values
+            if (currentEnumValues != null) Collections.sort(currentEnumValues);
+            if (elasticsearchEnumValues != null) Collections.sort(elasticsearchEnumValues);
+
+            // compare two list of enum values
+            if (((currentEnumValues != null) && (!currentEnumValues.equals(elasticsearchEnumValues))) ||
+                ((currentEnumValues == null) && (elasticsearchEnumValues != null) && (elasticsearchEnumValues.size() > 0)))
+            {
+                // if not equal, create or update enums index in elastic search
+                BluefloodEnumRollup rollupWithEnumValues = createRollupWithEnumValues(currentEnumValues);
+                IMetric enumMetric = new PreaggregatedMetric(0, locator, null, rollupWithEnumValues);
+                try {
+                    discoveryIO.insertDiscovery(enumMetric);
+                }
+                catch (Exception e) {
+                    log.error(String.format("Exception writing enums index to elasticsearch for %s: %s", locator.toString(), e.getMessage()), e);
+                }
+            }
+        }
+    }
+
```

Comment:
nit: nested ifs. is it easy and more readable to extract some of them into their own methods?